File: leetcode/0098-validate-binary-search-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return helper(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
    public boolean helper(TreeNode root, long min, long max){
        if(root == null) return true;

        //this means that 
        if(root.val <= min || root.val >= max) return false;
        
        boolean leftValid = helper(root.left, min, root.val);
        boolean rightValid = helper(root.right, root.val, max);
        return leftValid && rightValid;
    }
}

File: leetcode/0098-validate-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValid(root, float('-inf'), float('inf'))
    
    def isValid(self, root: Optional[TreeNode], minVal: int, maxVal: int) -> bool:
        if root is None:
            return True
        
        if root.val <= minVal or root.val >= maxVal:
            return False
        
        return self.isValid(root.left, minVal, root.val) and self.isValid(root.right, root.val, maxVal)
